lib/SILGen/SILGenExpr.cpp

```diff
@@ -3782,7 +3783,69 @@ visitMagicIdentifierLiteralExpr(MagicIdentifierLiteralExpr *E, SGFContext C) {
   llvm_unreachable("Unhandled MagicIdentifierLiteralExpr in switch.");
 }
 
-RValue RValueEmitter::visitCollectionExpr(CollectionExpr *E, SGFContext C) {
+RValue RValueEmitter::visitArrayExpr(ArrayExpr *E, SGFContext C) {
+  FormalEvaluationScope Scope(SGF);
```

Comment:
Why does this need a formal evaluation scope? You aren't working with SILGenLValue stuff here.

Reply:
I see, maybe I'm doing something wrong, but this is required by:
    auto cleanup =
        SGF.enterDormantFormalAccessTemporaryCleanup(destAddr, loc, destTL);
Which I lifted from the part of the TupleShuffleExpr lowering logic that handled constructing varargs.

Reply:
I think you probably want a full scope here.

Reply:
Sounds good.
Just a few questions:
would that be: ArgumentScope? Scope does not look sufficient.

I noticed that emitApplyExpr(...) which would normally be called is wrapping everything in a FormalEvaluationScope:

CallEmission CallEmission::forApplyExpr(SILGenFunction &SGF, Expr *e) {
   // Set up writebacks for the call(s).
   FormalEvaluationScope writebacks(SGF);
...

Reply:
I appreciate your thorough critique as I'm a tad out of my depth here. I suppose what you're getting at is that because all the sub-exprs are RValues and it is evaluating to a RValue that I don't need any FormalScope or cleanups? I updated it to pass invalidCleanup here and it seems to still work (All the cleanups seemed to do things related to formal scopes), but I'd like to get your thoughts.

Also, I'm interested in learning more about SILGen. If you have some cleanup tasks related to this API standardization effort, I would be interested.